Modernize project infrastructure

mdomke opened the pull request:
* Use `hatch` as build tool and for environment handling
* Use `pytest` as test-runner
* Use PEP 621 compliant project metadata (`pyproject.toml`)
* Fix some linter errors and test failures along the way
* Use GitHub Actions for CI
* Remove support for deprecated Python versions (everything prior to 3.8)

## Future plans
* Switch to `ruff`/`hatch` as linter and formatter
* ~~Use `black` as formatter and enforce formatting with `precommit` rules~~
* Modernize the code with `pyupgrade` to remove relics from Python2 times e.g (`class Foo(object):`)

mdomke committed:
Use hatch as project management tool while switching to pyproject.toml

mdomke committed:
Update license to ISC

mdomke committed:
Expose mongodb port locally for easier testing

mdomke committed:
Remove unsupported Python versions

mdomke committed:
Fix tests for pymongo versions 4.6

mdomke committed:
Remove legacy version dance for unsupported Pythons

mdomke committed:
Fix pylint warnings

mdomke committed:
Fix flake8 error about ambiguous l variable

mdomke committed:
Move flake8 options to .flake8 config file

mdomke committed:
Add GitHub Actions test pipeline

mdomke committed:
Start MongoDB server in GHA workflow

mdomke committed:
Use codecov and nodejs actions

mdomke committed:
chore: Specify dependencies for cov env

mdomke committed:
chore: Update to setup-python action v5

mdomke committed:
chore: Add codecov upload token

mdomke committed:
chore: Fix links in README

pcorpet commented on Dockerfile:
```diff
@@ -1,15 +1,14 @@
 
 RUN eval "$(pyenv init -)" && eval "$(pyenv virtualenv-init -)"
-RUN pyenv install 3.6.15
-RUN pyenv install 3.7.9
```
Make sure to update the pyenv local line below

pcorpet replied:
Please consider adding 3.11 and 3.12 as well. And also probably a more recent pypy3

pcorpet commented on docker-compose.yml:
```diff
@@ -19,3 +19,6 @@ services:
     # the value in mongomock/__init__.py
     image: mongo:5.0.5
+    ports:
+      - 27017:27017
```
This poses a security threat: you're opening a port on the host. If you want to check locally, I'd recommend using
```yml
ports:
  - localhost:27017:27017
```

mdomke replied:
`localhost` doesn't work, but I can use `127.0.0.1` instead. I'm not sure about the attack vector however

pcorpet commented on LICENSE:
```diff
@@ -1,24 +1,15 @@
-Copyright (c) 2012, Rotem Yaari <vmalloc@gmail.com>
```
You'll need @vmalloc's approval on that one.

mdomke replied:
Yes, you're right. I'll try to contact him.

mdomke replied:
I haven't heard back so far from @vmalloc. I will change the license back for the time being

pcorpet commented on pyproject.toml:
```diff
@@ -0,0 +1,77 @@
+authors = [
+    { name = "Rotem Yaari", email = "vmalloc@gmail.com" },
+    { name = "Martin Domke", email = "mail@martindomke.net" },
+]
```
Please add myself in there `{ name = "Pascal Corpet", email = "pascal@corpet.net" },`

pcorpet commented on tox.ini:
```diff
@@ -1,13 +1,11 @@
-  py37: python3.7
   py38: python3.8
   py39: python3.9
   py310: python3.10
```
Please add python 3.11 and python 3.12 (or remove them from the other places).

mdomke replied:
The point is to eventually replace `tox` with `hatch`, but I can certainly add the additional Python versions for now

pcorpet reviewed (changes requested):
Thanks a lot for this big cleanup. In general I'm more in favor of having multiple small PRs. If you can, please send some fixes in the python code separately from the changes of BUILD and separately from the changes in CI.

If too much of a hastle, let's continue it here although the review my be longer.

mdomke committed:
chore: Use hatch in docker-compose

mdomke committed:
Add Pascal Corpet as author

mdomke committed:
chore: Add Python 3.11/12 to tox

mdomke commented:
Hej @pcorpet. I agree that pull-requests should be kept related to a specific change. If it is okay with you, I would like to avoid going the extra mile and slice this pull-request into several smaller ones. What do you think?

mdomke committed:
Switch back to original BSD license

I couldn't get any answer from Rotem Yaari so that
I had to revert to the original license text

mdomke commented:
Also, @pcorpet, could you add the PyPI API token as a GitLab secret so that we can also switch to GitHub Actions for the deployment?

vmalloc commented:
@mdomke @pcorpet I just noticed this in the pile of notifications... Can you please explain what is needed here?

mdomke commented:
@vmalloc Originally I wanted to update the license to the [ISC license](https://choosealicense.com/licenses/isc/) and change the copyright owners to "Mongomock Contributors". 

> The ISC license is functionally equivalent to the [BSD 2-Clause](https://choosealicense.com/licenses/bsd-2-clause/) and [MIT](https://choosealicense.com/licenses/mit/) licenses, removing some language that is no longer necessary.

@pcorpet rightfully made me aware that for this kind of change, we would need your approval and that's why I was reaching out to you. In the meantime, I switched (commit: [f44fc3e](https://github.com/mongomock/mongomock/pull/876/commits/f44fc3e66f3319680e383b53daeff36e2683a302)) the license back to the original BSD 2-Clause style to not block this PR.

vmalloc commented:
So I'm good with the license change to ISC as long as my name is still somewhere in the authors list

mdomke commented:
It is and always will be 😄  

See: https://github.com/mongomock/mongomock/pull/876/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R11-R15

mdomke committed:
chore: Switch to hatch 1.12 test command

mdomke committed:
chore: Use hatch install action consistently

mdomke committed:
chore: Revert to older pylint version for now

mdomke committed:
Revert "Switch back to original BSD license"

This reverts commit f44fc3e66f3319680e383b53daeff36e2683a302.

mdomke commented:
@pcorpet I would like to proceed with this pull-request and if it is just only for being able to use GitHub Actions. After clarifying the license change with @vmalloc I also included the ISC license again.

mdomke committed:
Update docker image and compose definitions

mdomke committed:
Adapt README to hatch commands

mdomke committed:
Add GitHub Actions publish pipeline
